doc/developer/design/20230216_role_based_access_control.md

```diff
@@ -0,0 +1,384 @@
+- `REASSIGN OWNED BY <old_role> TO <new_role>`
+    - Transfers ownership of all objects owned by `<old_role>` to `<new_role>`.
+    - Can only be run by a member of `<old_role>` and `<new_role>` or a superuser.
+    - Requires `CREATE` privilege on all schemas and databases where all objects reside.
+        - TODO: This isn't explicitly stated in th docs, but would make sense base on the `ALTER` privileges. I will
+          double-check this.
+    - In PostgreSQL, this only affects the current database. We will diverge here and have it affect all databases.
+- `DROP OWNED BY <name> [ CASCADE | RESTRICT]`
+    - Drops all objects owned by `<name>`.
+    - Requires membership of `<name>`.
+    - In PostgreSQL, this only affects the current database. We will diverge here and have it affect all databases.
+    - Revokes all privileges granted to `<name>`.
+    - Does not drop owned databases or clusters.
+    - Default is `RESTRICT`.
+
+We will update `DROP ROLE` so that roles cannot be dropped until it meets the following criteria:
+
+- No objects are owned by the role.
+- The role contains no privileges.
+
+We will update `DROP <object>` so that it revokes all privileges on `<object>`.
+
+#### Implementation Details
+
+- Privileges will be stored in the catalog.
+- Privileges will be checked before operations in the sequencer.
+
+#### Out of Scope for Project
+
+- `CURRENT_ROLE`, `CURRENT_USER`, `SESSION_USER`, aliases in `GRANT`, `REVOKE`, `ALTER`, `REASSIGN OWNED`
+  and `DROP OWNED`.
+- `GRANTED BY` option for `GRANT` and `REVOKE`.
+- `WITH GRANT OPTION` in `GRANT`.
+- `GRANT OPTION FOR` in `REVOKE`.
+- The following privileges:
+    - `TRUNCATE`
+    - `REFERENCES`
+    - `TRIGGER`
+    - `CONNECT`
+    - `TEMPORARY`
+    - `EXECUTE`
+    - `SET`
+    - `ALTER SYSTEM`
+- The following object types:
+    - `DOMAIN`
+    - `FUNCTION` or `PROCEDURE`
+    - `FOREIGN DATA WRAPPER`
+    - `FOREIGN SERVER`
+    - `LANGUAGE`
+    - `LARGE OBJECT`
+    - `PARAMETER`
+    - `SEQUENCE`
+    - `Table column`
+    - `TABLESPACE`
+    - `TYPE`
+- Adding the necessary pg views to support all role based `psql` meta-commands.
+
+## Alternatives
+
+- PostgreSQL breaks from the SQL standard in a couple of places in their RBAC implementation. We may want to consider a
+  design that is more in line with the SQL standard than PostgreSQL.
+
+## Open questions
+
+- PostgreSQL grants certain privileges on certain object types to all roles. None of those overlap with the set of
+  objects and privileges we have, but do we want to do something similar?
+- Do we want different `SELECT` privileges based on if a new dataflow will be spun up or if we can use an existing one?
+    - Pros: We can differentiate between using existing compute resources vs creating new ones when reading.
+    - Cons: Users (and the database) are unable to determine if they're allowed to execute a read until after that read
```

Comment:
My take is we should avoid this for now. It's a pretty darn specific privilege, and it's one that we can always introduce later (`FAST_PATH_ONLY_USAGE` or something). Would be good to suss out how much of a need for this there actually is.

I'm not sure that it will be common for e.g. one team to say "this is our cluster, feel free to issue fast queries but absolutely no slow queries"—especially since enough fast queries can probably meaningfully impact cluster performance. Would rather encourage folks to create materialized views and say "if you want to query our views, turn on your own cluster and build your own indexes."

For example, an alternative to explore here is a _session parameter_ that prevents you from running non-fast-path peeks. I can imagine you might want to opt into this for some applications where you want to be really sure that you don't accidentally have a regression (either in your application code or in Materialize's optimizer) that causes queries to suddenly start taking the slow path and crash the cluster.